proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,553 @@
+identify, and display an exception, but lacks any generic facility for encoding
+relevant metadata about the exception.
+
+As a first step towards attaching backtraces to exceptions, we propose to extend
+``SomeException`` to carry an additional "context" value, consisting of a collection
+of dynamically-typed annotations. The dynamic nature of these annotations
+allows the mechanism to handle not just backtraces but any arbitrary
+context-relevant value which may be inspected by a handler for analysis and logging.
+To avoid breaking existing users of ``SomeException``, we introduce this
+context as an implicit parameter constraint:  ::
+
+    data SomeException where
+        SomeException :: forall a. (Exception a, ?exceptionContext :: ExceptionContext)
+                      => a -> SomeException
+
+    data ExceptionContext = ExceptionContext [SomeExceptionAnnotation]
+
+    -- | Contexts are merged by concatenation
+    instance Semigroup ExceptionContext
+    instance Monoid ExceptionContext
+
+Following the example of the ``Exception`` class, we propose that exception
+annotations implement a minimal typeclass which provides ``Typeable`` evidence
+and the ability to render the annotation to a ``String``: ::
+
+    data SomeExceptionAnnotation where
+        SomeExceptionAnnotation :: forall a. (ExceptionAnnotation a)
+                                => a -> SomeExceptionAnnotation
+
+    class Typeable a => ExceptionAnnotation a where
+        displayExceptionAnnotation :: a -> String
+        
+        default displayExceptionAnnotation :: (Show a) => a -> String
+        displayExceptionAnnotation = show
+
+To allow users to populate this new annotation field we propose that the
+``Exception`` class be extended with a new ``toExceptionWithContext`` method: ::
+
+    class (Typeable e, Show e) => Exception e where
+        -- These are unchanged:
+        toException            :: e -> SomeException
+        fromException          :: SomeException -> Maybe e
+        displayException       :: e -> String
+
+        -- This is new:
+        toExceptionWithContext :: e -> ExceptionContext -> SomeException
+        toExceptionWithContext e ?exceptionContext = SomeException e
+
+        -- toException is implemented in terms of toExceptionWithContext
+        toException e = toExceptionWithContext e mempty
+
+The ``Exception`` instance for ``SomeException`` would accumulate and display
+contexts: ::
+
+    instance Exception SomeException where
+        -- This is unchanged:
+        toException se = se
+        fromException = Just
+
+        -- toExceptionWithContext *adds* context to an existing SomeException:
+        toExceptionWithContext se ctxt = addExceptionContext ctxt se
+
+        -- displayException shows context after the exception itself:
+        displayException (SomeException e) =
+            displayException e ++ "\n" ++ displayExceptionContext ?exceptionContext
```

Comment:
Even if the exception context is empty, this will add an extra trailing newline. Perhaps make `displayExceptionContext` add this only if the context is non-empty?

Reply:
Quite right. I shall fix this.